jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/service/WildFlyCustomJtaPlatform.java

```diff
@@ -24,37 +24,69 @@
 import org.hibernate.engine.transaction.jta.platform.internal.JtaSynchronizationStrategy;
 
 /**
- * WildFlyCustomJtaPlatform
+ * WildFlyCustomJtaPlatform can obtain the JTA TransactionSynchronizationRegistry to be used by
+ * Hibernate ORM JPA + native applications.
+ * For JPA applications, we could of passed the TransactionSynchronizationRegistry into the
+ * constructor but Hibernate native apps wouldn't be able to do that, so this covers all app types.
  *
  * @author Scott Marlow
  */
-public class WildFlyCustomJtaPlatform extends JBossAppServerJtaPlatform {
+public class WildFlyCustomJtaPlatform extends JBossAppServerJtaPlatform implements JtaSynchronizationStrategy {
 
     private static final String TSR_NAME = "java:jboss/TransactionSynchronizationRegistry";
 
-    @Override
-    protected JtaSynchronizationStrategy getSynchronizationStrategy() {
-        final TransactionSynchronizationRegistry transactionSynchronizationRegistry =
-                locateTransactionSynchronizationRegistry();
+    // The 'transactionSynchronizationRegistry' used by JPA container managed applications,
+    // is reset every time the Transaction Manager service is restarted,
+    // as (application deployment) JPA persistence unit service depends on the TM service.
+    // For this reason, the static 'transactionSynchronizationRegistry' can be updated.
+    // Note that Hibernate native applications currently have to be (manually) restarted when the TM
+    // service restarts, as native applications do not have WildFly service dependencies set for them.
+    private static volatile TransactionSynchronizationRegistry transactionSynchronizationRegistry;
 
-        return new JtaSynchronizationStrategy() {
+    // JtaSynchronizationStrategy
+    @Override
+    public void registerSynchronization(Synchronization synchronization) {
+        transactionSynchronizationRegistry.registerInterposedSynchronization(synchronization);
+    }
 
+    // JtaSynchronizationStrategy
+    @Override
+    public boolean canRegisterSynchronization() {
+        return transactionSynchronizationRegistry.getTransactionStatus() == Status.STATUS_ACTIVE;
```

Comment:
This could produce an NPE.

Reply:
It requires the static `setTransactionSynchronizationRegistry()` to be invoked. That is at minimum racy. We should at least be guarding against an NPE IMO.

Reply:
It would be unexpected for it to be null but sure, I'll change it.

Reply:
Added null check.